app/javascript/lib/validation/expression-matrices-validation.js

```diff
@@ -0,0 +1,265 @@
+
+  dataObj.rowValues = dataObj.rowValues ? dataObj.rowValues : new Set()
+  dataObj.duplicateRowValues = dataObj.duplicateRowValues ? dataObj.duplicateRowValues : new Set()
+
+  if (!dataObj.rowValues.has(line)) {
+    dataObj.rowValues.add(line)
+  } else {
+    dataObj.duplicateRowValues.add(rawLine)
+  }
+  if (isLastLine && dataObj.duplicateRowValues.size > 0) {
+    const nameTxt = (dataObj.duplicateRowValues.size > 1) ? 'duplicates' : 'duplicate'
+    const dupString = [...dataObj.duplicateRowValues].slice(0, 10).join(', ')
+    const msg = `Row values must be unique within a file. ${dataObj.duplicateRowValues.size} ${nameTxt} found, including: ${dupString}`
+    issues.push(['error', 'duplicate:values-within-file', msg])
+  }
+  return issues
+}
+
+
+/**
+ * Verify dense matrix column numbers match header column numbers
+ */
+function validateColumnNumber(line, isLastLine, headers, lineNum, dataObj) {
+  const issues = []
+  dataObj.rowsWithIncorrectColumnNumbers = dataObj.rowsWithIncorrectColumnNumbers ? dataObj.rowsWithIncorrectColumnNumbers : []
+  // use the first header row to determine the correct number of columns all rows should have
+  const correctNumberOfColumns = headers[0].length
+
+  if (correctNumberOfColumns !== line.length) {
+    dataObj.rowsWithIncorrectColumnNumbers.push(lineNum)
+  }
+
+  const numBadRows = dataObj.rowsWithIncorrectColumnNumbers.length
+  if (isLastLine && numBadRows > 0) {
+    const rowText = numBadRows > 1 ? 'rows' : 'row'
+    const maxLinesToReport = 10
+    let notedBadRows = dataObj.rowsWithIncorrectColumnNumbers.slice(0, maxLinesToReport).join(', ')
+    if (numBadRows - maxLinesToReport > 0) {
+      notedBadRows += ` and ${numBadRows - maxLinesToReport} more rows`
+    }
+
+    const msg = `All rows must have the same number of columns. ` +
+      `Please ensure the number of columns for ${rowText}: ${notedBadRows}, ` +
+      `matches the file-header-specified number of ${correctNumberOfColumns} columns-per-row.`
+    issues.push(['error', 'format:mismatch-column-number', msg])
+  }
+
+  return issues
+}
+
+/**
+ * Verify sparse matrix has no blank lines
+ */
+function validateSparseNoBlankLines(line, isLastLine, lineNum, dataObj) {
+  const issues = []
+  dataObj.blankLineRows = dataObj.blankLineRows ? dataObj.blankLineRows : []
+
+  // if the line is empty, note it
+  if (!line.trim()) {
+    dataObj.blankLineRows.push(lineNum)
+  }
+
+  const numBadRows = dataObj.blankLineRows.length
+  if (isLastLine && numBadRows > 0) {
+    const rowText = numBadRows > 1 ? 'rows' : 'row'
+    const maxLinesToReport = 10
+    let notedBadRows = dataObj.blankLineRows.slice(0, maxLinesToReport).join(', ')
+    if (numBadRows - maxLinesToReport > 0) {
+      notedBadRows += ` and ${numBadRows - maxLinesToReport} more rows`
+    }
+
+    const msg = `Please ensure there are no blank rows in the file. ` +
+      `Remove or replace the following ${rowText}: ${notedBadRows}, `
```

Comment:
Blocker: your screenshot in response to Devon shows a typo -- a sentence ending with a comma.  I think this would fix it.

```suggestion
      `Remove or replace the following ${rowText}: ${notedBadRows}.`
```